app/src/main/java/io/appium/settings/RecorderThread.java

```diff
@@ -0,0 +1,367 @@
+            @Override
+            public void run() {
+                Thread.currentThread().setPriority(Thread.MAX_PRIORITY);
+                try {
+                    audioRecord.startRecording();
+                } catch (Exception e) {
+                    asyncError = true;
+                    e.printStackTrace();
+                    return;
+                }
+                try {
+                    while (!audioStopped) {
+                        int index = audioEncoder.dequeueInputBuffer(10000);
+                        if (index < 0) {
+                            continue;
+                        }
+                        ByteBuffer inputBuffer = audioEncoder.getInputBuffer(index);
+                        if (inputBuffer == null) {
+                            if (!stopped) {
+                                asyncError = true;
+                            }
+                            return;
+                        }
+                        inputBuffer.clear();
+                        int read = audioRecord.read(inputBuffer, inputBuffer.capacity());
+                        if (read < 0) {
+                            if (!stopped) {
+                                asyncError = true;
+                            }
+                            break;
+                        }
+                        audioEncoder.queueInputBuffer(index, 0, read,
+                                getPresentationTimeUs(), 0);
+                    }
+                } catch (Exception e) {
+                    if (!stopped) {
+                        Log.e(TAG, "Audio error", e);
+                        asyncError = true;
+                        e.printStackTrace();
+                    }
+                } finally {
+                    audioRecord.stop();
+                    audioRecord.release();
+                    audioRecord = null;
+                }
+            }
+        });
+        audioRecordThread.start();
+    }
+
+    private long getPresentationTimeUs() {
+        if (!startTimestampInitialized) {
+            startTimestampUs = System.nanoTime() / 1000;
+            startTimestampInitialized = true;
+        }
+        return (System.nanoTime() / 1000 - startTimestampUs) / timeLapse;
+    }
+
+    private void startMuxerIfSetUp() {
+        if (audioTrackIndex >= 0 && videoTrackIndex >= 0) {
+            muxer.start();
+            muxerStarted = true;
+        }
+    }
+
+    @SuppressLint("NewApi")
+    @Override
+    public void run() {
+        try {
+            setupVideoCodec();
+            setupVirtualDisplay();
+
+            setupAudioCodec();
+            setupAudioRecord();
+
+            muxer = new MediaMuxer(this.outputFilePath,
+                    MediaMuxer.OutputFormat.MUXER_OUTPUT_MPEG_4);
+
+            startAudioRecord();
+
+            MediaCodec.BufferInfo bufferInfo = new MediaCodec.BufferInfo();
+            long lastAudioTimestampUs = -1;
+
+            while (!stopped && !asyncError) {
```

Comment:
is it possible to extract audio and video recording to separate private methods to make this one easier to read?

Reply:
done, https://github.com/sirmordred/io.appium.settings/commit/396e861b2d723258f833c87222448d1c52a10bc6